[TypeProf][InstrPGO] Introduce raw and instr profile format change for type profiling.

mingmingl-llvm opened the pull request:
* Raw profile format
   - Header: records the byte size of compressed vtable names, and the number of profiled vtable entries (call it `VTableProfData`). Header also records padded bytes of each section.
   - Payload: adds a section for compressed vtable names, and a section to store `VTableProfData`. Both sections are padded so the size is a multiple of 8.
* Indexed profile format
  - Header: records the byte offset of compressed vtable names.
  - Payload: adds a section to store compressed vtable names. This section is used by `llvm-profdata` to show the list of vtables profiled for an instrumented site.
  
[The originally reviewed patch](https://github.com/llvm/llvm-project/pull/66825) will have profile reader/write change and llvm-profdata change.
- To ensure this PR has all the necessary profile format change along with profile version bump, created a copy of the originally reviewed patch in https://github.com/llvm/llvm-project/pull/80761. The copy doesn't have profile format change, but it has the set of tests which covers type profile generation, profile read and profile merge. Tests pass there.
  
rfc in https://discourse.llvm.org/t/rfc-dynamic-type-profiling-and-optimizations-in-llvm/74600

mingmingl-llvm committed:
[TypeProf]Introduce value type in raw profiles

mingmingl-llvm committed:
undo the changes that don't necessarily bundle with profile format
change, to get smaller PRs.

mingmingl-llvm committed:
fix the alignment of struct

mingmingl-llvm committed:
polish format

mingmingl-llvm committed:
fix format

minglotus-6 commented:
I have to ask for another round of review , after the original PR (https://github.com/llvm/llvm-project/pull/66825) received many many feedbacks. Thanks all for the time again!

david-xl commented on compiler-rt/include/profile/InstrProfData.inc:
```diff
@@ -94,6 +94,25 @@ INSTR_PROF_DATA(const uint32_t, llvm::Type::getInt32Ty(Ctx), NumBitmapBytes, \
+#endif
+INSTR_PROF_VTABLE_DATA(const uint64_t, llvm::Type::getInt64Ty(Ctx), \
+                       VTableNameHash, ConstantInt::get(llvm::Type::getInt64Ty(Ctx), \
+                     IndexedInstrProf::ComputeHash(PGOVTableName)))
```
indentation

minglotus-6 replied:
fixed.

david-xl commented on compiler-rt/lib/profile/InstrProfilingMerge.c:
```diff
@@ -107,6 +107,26 @@ static uintptr_t signextIfWin64(void *V) {
 
+static uint64_t
+getDistanceFromCounterToValueProf(const __llvm_profile_header *const Header) {
+  // Skip names section, vtable profile data section and vtable names section
```
Move comments outside function body.

minglotus-6 replied:
done.

david-xl commented on llvm/lib/ProfileData/InstrProfWriter.cpp:
```diff
@@ -455,12 +455,13 @@ Error InstrProfWriter::writeImpl(ProfOStream &OS) {
-  for (int I = 0; I < N - 4; I++)
+  // 'BinaryIdOffset', `TemporalProfTracesOffset` and `VTableNamesOffset`. We
+  // need to remember the offset of these fields to allow back patching later.
+  for (int I = 0; I < N - 5; I++)
```
define a macro "#define SKIPPED_FIELDS 5"

minglotus-6 replied:
Changed this to write out the first four fields and updated comment.

david-xl commented on llvm/lib/ProfileData/InstrProfWriter.cpp:
```diff
@@ -604,6 +608,35 @@ Error InstrProfWriter::writeImpl(ProfOStream &OS) {
 
+  // if version >= the version with vtable profile metadata.
+  uint64_t VTableNamesSectionStart = 0;
+  if (IndexedInstrProf::ProfVersion::CurrentVersion >= 12) {
```
Is the check needed? Instead add an assertion on the version.

minglotus-6 replied:
The check is verbose. I removed it.

mingmingl-llvm committed:
resolve review feedback

mingmingl-llvm committed:
add back one blank line in llvm/include/llvm/ProfileData/InstrProfData.inc

minglotus-6 commented:
thanks for the review! I'm now communicating with the stakeholders for coordination. Once there is a plan (hopefully as late as mid next week), I will merge this change.

modiking commented on compiler-rt/include/profile/InstrProfData.inc:
```diff
@@ -186,13 +207,28 @@ VALUE_PROF_FUNC_PARAM(uint32_t, CounterIndex, Type::getInt32Ty(Ctx))
+/* For virtual table address profiling, the address point of the virtual table
+ * (i.e., the address contained in objects pointing to a virtual table) are
+ * profiled. Note this may not be the address of the per C++ class virtual table
+ *  object (e.g., there might be an offset).
```
```suggestion
 * object (e.g., there might be an offset).
```

modiking commented on compiler-rt/include/profile/InstrProfData.inc:
```diff
@@ -186,13 +207,28 @@ VALUE_PROF_FUNC_PARAM(uint32_t, CounterIndex, Type::getInt32Ty(Ctx))
+ * 1. The (starting and ending) addresses of per C++ class virtual table objects.
+ * 2. The (compressed) virtual table object names.
+ * RawInstrProfReader converts profiled virtual table addresses to virtual table
+ *  objects' MD5 hash.
```
```suggestion
 * objects' MD5 hash.
```

modiking commented on compiler-rt/include/profile/InstrProfData.inc:
```diff
@@ -186,13 +207,28 @@ VALUE_PROF_FUNC_PARAM(uint32_t, CounterIndex, Type::getInt32Ty(Ctx))
+ *  objects' MD5 hash.
+ */
+VALUE_PROF_KIND(IPVK_VTableTarget, 2, "The address of the compatible vtable (i.e., "
+                                      "there is an offset from this address to per C++ "
```
```suggestion
                                      "there is an offset from this address to a C++ "
```

modiking commented on compiler-rt/include/profile/InstrProfData.inc:
```diff
@@ -186,13 +207,28 @@ VALUE_PROF_FUNC_PARAM(uint32_t, CounterIndex, Type::getInt32Ty(Ctx))
  * indexed with the kind value.
  */
 VALUE_PROF_KIND(IPVK_First, IPVK_IndirectCallTarget, "first")
-VALUE_PROF_KIND(IPVK_Last, IPVK_MemOPSize, "last")
```
Is IPVK_MemOPSize not needed anymore?

minglotus-6 replied:
Before the change, `IPVK_Last` has the value of `IPVK_MemOPSize`; it's intentional to update `IPVK_Last` to `IPVK_VTableTarget` such that value profiling code path handle the new type of value profile.

More specifically, the raw profile file contains memory dump of struct 
[`__llvm_profile_data`](https://github.com/llvm/llvm-project/blob/118a2a52fd465b7576c54bd102ee2e417a3b9a71/compiler-rt/lib/profile/InstrProfiling.h#L36), and we need the space for [`NumValueSites`](https://github.com/llvm/llvm-project/blob/118a2a52fd465b7576c54bd102ee2e417a3b9a71/compiler-rt/include/profile/InstrProfData.inc#L90) as a raw profile format change, and thereby the need to bump the raw profile version.
* `IPVK_Last` is used in other places besides raw profile format change. These code paths just work out (e.g., `pgo-instr-gen` pass will get an empty instrumented site for the new `IPVK_VTableTarget` kind in its [`ValueSite`](https://github.com/llvm/llvm-project/blob/118a2a52fd465b7576c54bd102ee2e417a3b9a71/llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp#L570-L587) but that won't cause correctness problems)

modiking replied:
Oh I see, this is just marking the start and end of the array not the actual definition of the profile types.

modiking commented on compiler-rt/lib/profile/InstrProfilingWriter.c:
```diff
@@ -273,18 +280,28 @@ lprofWriteDataImpl(ProfDataWriter *Writer, const __llvm_profile_data *DataBegin,
-      &PaddingBytesAfterBitmapBytes, &PaddingBytesAfterNames);
+      PaddingBytesAfterBitmapBytes, PaddingBytesAfterNames,
+      PaddingBytesAfterVTable, PaddingBytesAfterVNames;
+  if (__llvm_profile_get_padding_sizes_for_counters(
```
`__llvm_profile_get_padding_sizes_for_counters` return both `0` and `-1` on failure and this here is only checking for -1. Is that intentional?

If the tri-state is needed I would be very explicit on having an enum for `success`, `failure`, or `error`. Otherwise, would 2 states be enough?

minglotus-6 replied:
`__llvm_profile_get_padding_sizes_for_counters` returns -1 to alert the [incompatibility](https://github.com/llvm/llvm-project/pull/81691/files/0c9ed57bce7d29f27ab7562d85503edfa8132670#diff-ff206eae21de963bf60f9baca54d2c90e0dd89bf53732c236b019b4e9192bd20R163) and returns 0 otherwise.

-1 is chosen to signal errors to for consistency with existing code (e.g.,  [`InstrProfilingWriter.c`](https://github.com/llvm/llvm-project/blob/118a2a52fd465b7576c54bd102ee2e417a3b9a71/compiler-rt/lib/profile/InstrProfilingWriter.c) uses -1 to return error in many places, and [InstrProfilingFile.c:writeFile](https://github.com/llvm/llvm-project/blob/118a2a52fd465b7576c54bd102ee2e417a3b9a71/compiler-rt/lib/profile/InstrProfilingFile.c#L518) also returns -1). Non-zero value is handled [here](https://github.com/llvm/llvm-project/blob/118a2a52fd465b7576c54bd102ee2e417a3b9a71/compiler-rt/lib/profile/InstrProfilingFile.c#L605-L606) to emit warnings.

Anecdotally, `llvm_unreachable` is not a way to signal errors in release build with the default LLVM config.
 * https://llvm.org/docs/CodingStandards.html#assert-liberally mentions this _When assertions are disabled (i.e. in release builds), llvm_unreachable becomes a hint to compilers to skip generating code for this branch_ 
 * https://llvm.org/docs/CMake.html#llvm-related-variables mentions `DLLVM_UNREACHABLE_OPTIMIZE` is on by default and enables the hint

minglotus-6 replied:
Also, compiler-rt/lib/profile is in C (but not C++ yet afaik, see https://discourse.llvm.org/t/using-c-and-sanitizer-common-in-profile-runtime/58629) so we don't have `bool` type either.

modiking replied:
I see, even though there's precedent for using -1 I'm personally not a big fan because it's an uncommon pattern. Also I've usually found for many llvm functions that a `bool` of `true` indicates success and `false` failure which causes the opposite pattern (an explicit `bool` type isn't necessary for the same behavior).

If `unreachable` was considered as a way to signal error can error out immediately at the fail site? Naively a direct termination, if there's an error handle around invoking that would be good.

minglotus-6 replied:
To clarify a little bit, we don't want to use `llvm_unreachable` to signal errors at least for release builds.

> Also I've usually found for many llvm functions that a bool of true indicates success and false failure which causes the opposite pattern (an explicit bool type isn't necessary for the same behavior).

I agree that using enums has better readability in general. However, given the callers on the chain of the current function regards non-zero (concretely -1) as error and 0 as ok, it's a larger change (that usually goes to a separate PR) to improve the readability using enums. The `true` and `false` interpretation in llvm depends a lot on the caller context based on my experience.

modiking commented on llvm/lib/ProfileData/InstrProfReader.cpp:
```diff
@@ -1260,6 +1283,19 @@ Error IndexedInstrProfReader::readHeader() {
+        support::endian::readNext<uint64_t, llvm::endianness::little,
+                                  unaligned>(Ptr);
+
+    VTableNamePtr = (const char *)Ptr;
```
Should there be a check that this is a valid pointer before dereferencing it against buffer end like above?

minglotus-6 replied:
this makes sense, done by returning 'truncated' error.

modiking commented:
Also the CI clang-format is complaining so make sure the re-run that on your changes

mingmingl-llvm committed:
Apply suggestions from code review

Co-authored-by: modiking <modiking213@gmail.com>

mingmingl-llvm committed:
return truncated error if any

minglotus-6 commented:
> Also the CI clang-format is complaining so make sure the re-run that on your changes

clang-format formats existing code (specifically C++ macros) in `InstrProfData.inc` pretty aggressively and the indentations making code hard to read. 

Taking https://github.com/llvm/llvm-project/actions/runs/7921397694/job/21626608711?pr=81691 as an example, reformatted code between line 210 to line 267 are hard to read, same for code between line 284 to line 404. 

Given clang-format only complains about two `InstrProfData.inc`, I'm inclined not to run `git clang-format` to update these two files.

snehasish commented on llvm/lib/ProfileData/InstrProfWriter.cpp:
```diff
@@ -455,12 +455,12 @@ Error InstrProfWriter::writeImpl(ProfOStream &OS) {
-  for (int I = 0; I < N - 4; I++)
+  // Only write out the first four fields. We need to remember the offset of the
+  // remaining fields to allow back patching later.
+  for (int I = 0; I < 4; I++)
```
Should this be 5 - i.e. `Magic, Version, Unused, HashType, HashOffset`? If so can you also make sure we have a test to catch this.

david-xl replied:
should be 4 -- hashoffset field need patch up.

My original review comments was to introduce a macro instead of using a constant literal. The original macro name i suggested is not great. Perhaps do this:

#define NUM_FIELDS_PATCHING 5

// only write up first fields that do not need patching
for (int I = 0; I < N - NUM_FIELDS_PATCHING; I++) {
   ...
}

minglotus-6 replied:
thanks for the comment @snehasish! While it turns out `compiler-rt/test/` has the test coverage we want [1], a change like this warrants a higher degree of caution.

@david-xl I wonder what's the reason we want to keep 'N - constant' here? I'm open to make the change, yet felt the current way saves a bump to `constant` if more fields are appended in the header struct later.
* We need patching since profile summary section comes right after header (and before hashing table). To maintain backward compatibility of the indexed profiles, we don't want to delete existing header fields or adjust their order.


[1] Change this to 5 causes failures in `compiler-rt/test`

david-xl replied:
Your reasoning makes sense: the invariant is actually the prefix part, not the fields needs patching. Your current version LGTM.

modiking commented:
> > Also the CI clang-format is complaining so make sure the re-run that on your changes
> 
> clang-format formats existing code (specifically C++ macros) in `InstrProfData.inc` pretty aggressively and the indentations making code hard to read.
> 
> Taking https://github.com/llvm/llvm-project/actions/runs/7921397694/job/21626608711?pr=81691 as an example, reformatted code between `-/+` line 210 to line 267 are hard to read, same for code between line 284 to line 404, code between 780 to 970, etc
> 
> Given clang-format only complains about two `InstrProfData.inc`, I'm inclined not to run `git clang-format` to update these two files.

If that's the case we could ignore these files from `clang-format` via a `.clang-format-ignore` file if you're inclined to make that change.

modiking reviewed (approved):
LGTM

snehasish commented on compiler-rt/lib/profile/InstrProfilingBuffer.c:
```diff
@@ -191,7 +238,10 @@ COMPILER_RT_VISIBILITY int __llvm_profile_write_buffer_internal(
+  // Set virtual table arguments to NULL since they are not supported yet.
+  return lprofWriteDataImpl(
+      &BufferWriter, DataBegin, DataEnd, CountersBegin, CountersEnd,
+      BitmapBegin, BitmapEnd, 0 /* VPDataReader */, NamesBegin, NamesEnd,
```
If you specify the name before the argument like this: `/*VPDataReader=*/0` then clang-tidy bugprone [1] checks will catch it.

Another alternative would be to use options structs and named member initialization but that's more work (and some of the other methods here should be converted too).

[1] https://clang.llvm.org/extra/clang-tidy/checks/bugprone/argument-comment.html#bugprone-argument-comment

snehasish commented on llvm/lib/ProfileData/InstrProfWriter.cpp:
```diff
@@ -604,6 +607,31 @@ Error InstrProfWriter::writeImpl(ProfOStream &OS) {
+  // profile format change in place for version 12.
+  // TODO: Store the list of vtable names in InstrProfWriter and use the
+  // real compressed name.
+  std::string CompressedVTableNames;
```
Prefer using a dummy string, a lot of the code below will behave differently e.g. the for loop and the padding loop wont execute at all as far as I can tell.

minglotus-6 commented:
> > > Also the CI clang-format is complaining so make sure the re-run that on your changes
> > 
> > 
> > clang-format formats existing code (specifically C++ macros) in `InstrProfData.inc` pretty aggressively and the indentations making code hard to read.
> > Taking https://github.com/llvm/llvm-project/actions/runs/7921397694/job/21626608711?pr=81691 as an example, reformatted code between `-/+` line 210 to line 267 are hard to read, same for code between line 284 to line 404, code between 780 to 970, etc
> > Given clang-format only complains about two `InstrProfData.inc`, I'm inclined not to run `git clang-format` to update these two files.
> 
> If that's the case we could ignore these files from `clang-format` via a `.clang-format-ignore` file if you're inclined to make that change.

Thanks for the suggestions! After taking a look, `.clang-format-ignore` isn't used in current repo while `// clang-format off:<reason>` is used (e.g. at 7 different places in this [file](https://github.com/llvm/llvm-project/blob/450462cbaceddf57812ce15b5135b17f65a77654/clang/lib/Analysis/UnsafeBufferUsage.cpp)). I could make possibly add a few `// clang-format off` and `// clang-format on` around c++ macros in an nfc pull request.

Meanwhile, I'll send out a follow up change to update https://llvm.org/docs/InstrProfileFormat.html

snehasish commented on compiler-rt/include/profile/InstrProfData.inc:
```diff
@@ -94,6 +94,25 @@ INSTR_PROF_DATA(const uint32_t, llvm::Type::getInt32Ty(Ctx), NumBitmapBytes, \
+INSTR_PROF_VTABLE_DATA(const uint64_t, llvm::Type::getInt64Ty(Ctx), \
+                       VTableNameHash, ConstantInt::get(llvm::Type::getInt64Ty(Ctx), \
+                       IndexedInstrProf::ComputeHash(PGOVTableName)))
+INSTR_PROF_VTABLE_DATA(const IntPtrT, llvm::Type::getInt8PtrTy(Ctx), \
```
Diffing the contents here with the file in `llvm/ProfileData/` showed me that the latter uses `llvm::PointerType::getUnqual(Ctx)`. Can you sync them up and make the diff the files to ensure there are no inconsistencies?

minglotus-6 replied:
also sent out an [nfc](https://github.com/llvm/llvm-project/pull/82434) to clean up the use of `getInt8PtrTy`

snehasish commented on compiler-rt/lib/profile/InstrProfilingBuffer.c:
```diff
@@ -71,6 +75,19 @@ uint64_t __llvm_profile_get_data_size(const __llvm_profile_data *Begin,
+uint64_t __llvm_profile_get_num_vtable(const VTableProfData *Begin,
+                                       const VTableProfData *End) {
+  intptr_t EndI = (intptr_t)End, BeginI = (intptr_t)Begin;
+  return (EndI + sizeof(VTableProfData) - 1 - BeginI) / sizeof(VTableProfData);
```
I don't follow the math for the numerator. Can you add a comment to explain?

minglotus-6 replied:
Originally I implemented `__llvm_profile_get_num_vtable` based on `__llvm_profile_get_num_data`, and the question is around whether/why `sizeof(C struct) - 1` is necessary in the numerator.

From offline chat, the conclusion (based on reverse reasoning) is that `sizeof(C struct) - 1` is  necessary iff `[Begin, End]` is an inclusive range.
* For example, the function is expected to return 1 but would return 0 if caller passes `<Addr, Addr + sizeof(C-struct) -1>` as the pair of `<Begin, End>`  

From inspecting the elf sections of `__llvm_prof_cnts` in a simple instrumented hello-world program and looking at pointer values in a debugger, `End` is one byte past the inclusive range (i.e. caller passes `[Begin, End)` as function argument).

So it makes no difference (at least for elf) if we remove `sizeof(C struct) - 1 ` from the numerator. I added a comment in `__llvm_profile_get_num_data` to record the conclusion (but keep existing code as it is) and removed `sizeof()-1` in the new code.

minglotus-6 replied:
Or rather, I wonder if I shall make `__llvm_profile_get_num_vtable` and `__llvm_profile_get_num_data` consistent (both should use `sizeof(c struct) - 1` or both should not use them).

snehasish replied:
I think new code should not include it just because it was there before. If you update the old code and it breaks something post-commit then reverting this patch will be painful. If you want to keep things consistent, I would split out the change to the old code as a small patch and submit it soon (and monitor for any breakage - though I believe it is unlikely to happen).

minglotus-6 replied:
sounds good. I'll send out separate patches to update existing code (`__llvm_profile_get_num_data`). Before that I might dig and validate whether non-elf platforms also guarantee caller passes `[Begin, End)` (not `[Begin, End]`).

mingmingl-llvm committed:
resolve review fee feedback

minglotus-6 commented on compiler-rt/lib/profile/InstrProfilingBuffer.c:
```diff
@@ -71,6 +84,26 @@ uint64_t __llvm_profile_get_data_size(const __llvm_profile_data *Begin,
+COMPILER_RT_VISIBILITY
+uint64_t __llvm_profile_get_vtable_section_size(const VTableProfData *Begin,
+                                                const VTableProfData *End) {
+  return __llvm_profile_get_num_vtable(Begin, End) * sizeof(VTableProfData);
```
I updated both `__llvm_profile_get_num_vtable` and `__llvm_profile_get_vtable_section_size` to get rid of `sizeof(C-struct)`.

https://gcc.godbolt.org/z/MWaG6cEPn shows the generated codes before and after are equivalent (as long as `((intptr_t)End-(intptr_t)Begin) % sizeof(C-struct)==0` is true)

mingmingl-llvm committed:
resolve review feedback

mingmingl-llvm committed:
Merge branch 'main' into users/minglotus-6/typeprofrawformat

mingmingl-llvm committed:
restore __llvm_profile_get_num_vtable, and swap the order of two header fields in raw profile header

snehasish reviewed (approved):
lgtm

mingmingl-llvm committed:
llvm/test/tools/llvm-profdata/nocompress.test fails because its raw
profile version is outdated.
- update llvm/test/tools/llvm-profdata/Inputs/compressed.profraw along with raw profile version bump
- The test requires '!zlib', so I generate raw profile with DLLVM_ENABLE_ZLIB=1 and run the
  test with DLLVM_ENABLE_ZLIB=0.

minglotus-6 commented:
Going to run another 'git merge main' as a way to re-run pre-commit CI, given the current [Windows failure](https://buildkite.com/llvm-project/github-pull-requests/builds/40770#018dcde9-0609-49c9-8825-1217749b0bdc) doesn't seem relevant.
* Per https://discourse.llvm.org/t/how-to-rerun-buildkite-github-pull-requests/73402/6, push a new commit is a workaround to re-run pre-commit CI.

mingmingl-llvm committed:
Merge branch 'main' into users/minglotus-6/typeprofrawformat

minglotus-6 commented:
Hi,
 I reverted the change yesterday as the build of profiling runtime broke for non-ELF cases. I [restored the original branch ](https://github.com/llvm/llvm-project/compare/main...users/minglotus-6/typeprofrawformat) with all commit histories and appended one [commit](https://github.com/llvm/llvm-project/commit/a825d2a4ec00f07772a373091a702f149c3b0c34) on top of that and now waiting for Windows machine availability for a real test (MacOS test passed). Will reland the change after the Windows test.

minglotus-6 commented:
> Hi, I reverted the change yesterday as the build of profiling runtime broke for non-ELF cases. I [restored the original branch ](https://github.com/llvm/llvm-project/compare/main...users/minglotus-6/typeprofrawformat) with all commit histories and appended one [commit](https://github.com/llvm/llvm-project/commit/a825d2a4ec00f07772a373091a702f149c3b0c34) on top of that. Now waiting for Windows machine availability for a real test (MacOS test passed). Will send out a PR for review to reland the change after testing the change on Windows.

Actually I realized only Windows linker magic needs real machine for testing. To pass the build, the functions could return `NULL`. Even on ELF, those pointers are NULL (i.e., no lowering of data in the profile format change).

So sent out https://github.com/llvm/llvm-project/pull/82711
